### PR TITLE
correctly mask overlapping regions

### PR DESCRIPTION
**Pull request by mathause**
- [x] see #228, #265
 - [x] Tests added
 - [x] Passes `isort . && black . && flake8`
 - [x] Fully documented, including `whats-new.rst`

I thought about doing #265 but then I ended up coding #228... This is a "simple" and "inefficient" (for loop) implementation compared to what I proposed in #228. Still it is much more convenient as the original workaround. Although it technically fixes the issue in #228 I won't close it because I intend to improve on it later. Maybe wait for shapely 2.0 (and deprecate the pygeos masking) first.

---

- is `overlap` a good name for the keyword? I am not very keen to change it now but here are some suggestions:
  - `regionmask.Regions(..., overlap=True)`
  - `regionmask.Regions(..., overlaps=True)`
  - `regionmask.Regions(..., overlapping=True)`
  - others?

(The only thing I don't particularly like about `overlap` is when using it in a conditional, e.g. `if region.overlap:`. The other two would be better there but sound strange/ are too long IMHO.)

- there is currently no warning on `overlap=None` - should there be or can that wait?

cc @larsbuntemeyer, @ruthlorenz, @aaronspring

### CODE REVIEW

**Comment by aaronspring** [comment]
what do you not link in `if region.overlap`? if it should read like english then rather go for `overlaps`, but for the function call `overlap` sounds more normal to me

**Comment by larsbuntemeyer** [comment]
> 

i agree, keyword `overlap` sounds more normal!

File: regionmask/core/regions.py [hunk 1]
```diff
@@ -46,6 +47,16 @@ def __init__(
             Name of the collection of regions. Default: "unnamed"
         source : string, optional
             Source of the region definitions. Default: "".
+        overlap : bool, default: False
+            Indicates if (some of) the regions overlap. If True ``mask_3D`` will ensure
+            overlapping regions are correctly assigned grid points while ``mask`` will
```
Comment (mathause) [inline regionmask/core/regions.py:52]:
```suggestion
            overlapping regions are correctly assigned to grid points while ``mask`` will
```

File: regionmask/core/regions.py [hunk 2]
```diff
@@ -281,6 +293,12 @@ def mask(
         wrap_lon=None,
     ):
 
+        if self.overlap:
+            raise ValueError(
+                "Creating a 2D mask with overlap regions yields wrong results. "
```
Comment (mathause) [inline regionmask/core/regions.py:298]:
```suggestion
                "Creating a 2D mask with overlapping regions yields wrong results. "
```

**Comment by mathause** [comment]
Thanks for the feedback. I updated the docs to mention the default behavior. I think I will merge this soonish unless you have more comments?

**Comment by mathause** [comment]
Ok let's do this.

**Pull request merged**